curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -1382,6 +1384,33 @@ CURVEFS_ERROR DataCache::Flush(uint64_t inodeId, bool force) {
         VLOG(9) << "start datacache flush, chunkId:" << chunkId
                 << ",Len:" << tmpLen << ",blockPos:" << blockPos
                 << ",blockIndex:" << blockIndex;
+        // split req, init pendingReq
+        uint64_t ReqCounts = 0;
+        while (tmpLen > 0) {
+            if (blockPos + tmpLen > blockSize) {
+                n = blockSize - blockPos;
+            } else {
+                n = tmpLen;
+            }
+            tmpLen -= n;
+            blockPos = (blockPos + n) % blockSize;
+            ReqCounts++;
+        }
+        pendingReq.fetch_add(ReqCounts, std::memory_order_seq_cst);
+        VLOG(9) << "pendingReq is: " << pendingReq;
+        // init again
+        tmpLen = len_;
+        blockPos = chunkPos_ % blockSize;
+        PutObjectAsyncCallBack cb =
+                [&](const std::shared_ptr<PutObjectAsyncContext> &context) {
+                pendingReq.fetch_sub(1, std::memory_order_seq_cst);
+                if (!pendingReq.load(std::memory_order_seq_cst)) {
```

Comment:
merge fetch_sub and load into one statement, otherwise more than one thread can see pendingReq == 0
```cpp
if (pendingReq.fetch_sub(1) == 1) {
     cond.Signal();
}
```

Reply:
BTW, you should handle upload failed

Reply:
> BTW, you should handle upload failed

handle in S3Adapter::PutObjectAsync